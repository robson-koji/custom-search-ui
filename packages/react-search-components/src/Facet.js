import PropTypes from "prop-types";
import React from "react";

import { FacetOption } from "./types";

function Facet({ name, onRemove, onSelect, options, value }) {
  return (
    <div className="facet">
      <div>
        <div className="facet__title">{name}</div>
        <ul className="facet__list">
          {value && (
            <li className="facet__selected">
              {value}{" "}
              <span className="facet__remove">
                (
                <a
                  onClick={clickEvent => onRemove({ clickEvent, value })}
                  href="/"
                >
                  Remove
                </a>
                )
              </span>
            </li>
          )}
          {!value &&
            options.map(option => (
              <li className="facet__item" key={option.value}>
                <a
                  className="facet__link"
                  href="/"
                  onClick={clickEvent =>
                    onSelect({ clickEvent, value: option.value })
                  }
                >
                  {option.value}
                </a>{" "}
                <span className="facet__count">{option.count}</span>
              </li>
            ))}
        </ul>
      </div>
    </div>
  );
}

Facet.propTypes = {
  name: PropTypes.string.isRequired,
  onRemove: PropTypes.func.isRequired,
  onSelect: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(FacetOption).isRequired,
  value: PropTypes.string
};

export default Facet;
