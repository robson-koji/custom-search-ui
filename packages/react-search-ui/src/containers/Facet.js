import PropTypes from "prop-types";
import { Component } from "react";

import { FacetDetail, Filter } from "../types";
import { withSearch } from "..";

function findFacetValueInFilters(name, filters) {
  const filter = filters.find(f => Object.keys(f)[0] === name);
  if (!filter) return;
  return Object.values(filter)[0];
}

export class FacetContainer extends Component {
  static propTypes = {
    // Props
    render: PropTypes.func.isRequired,
    field: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    show: PropTypes.number,
    // State
    filters: PropTypes.arrayOf(Filter).isRequired,
    facets: PropTypes.objectOf(FacetDetail).isRequired,
    // Actions
    addFilter: PropTypes.func.isRequired,
    removeFilter: PropTypes.func.isRequired,
    setFilter: PropTypes.func.isRequired
  };

  constructor(props) {
    super();
    this.state = {
      more: props.show || 5
    };
  }

  handleClickMore = () => {
    this.setState(({ more }) => ({
      more: more + 10
    }));
  };

  render() {
    const { more } = this.state;
    const {
      render,
      field,
      label,
      addFilter,
      filters,
      facets,
      removeFilter,
      setFilter
    } = this.props;
    const facetValues = facets[field];
    if (!facetValues) return null;

    const options = facets[field][0].data;
    const selectedValues = findFacetValueInFilters(field, filters) || [];
    if (!options.length && !selectedValues.length) return null;

    return render({
      label: label,
      onMoreClick: this.handleClickMore,
      onRemove: value => {
        removeFilter(field, value);
      },
      onChange: value => {
        setFilter(field, value);
      },
      onSelect: value => {
        addFilter(field, value);
      },
      options: options.slice(0, more),
      showMore: options.length > more,
      values: selectedValues
    });
  }
}

export default withSearch(FacetContainer);
