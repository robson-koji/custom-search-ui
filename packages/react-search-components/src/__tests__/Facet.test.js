import React from "react";
import Facet from "../Facet";
import { shallow } from "enzyme";

const requiredProps = {
  name: "Facet",
  onRemove: () => {},
  onSelect: () => {},
  options: [{ value: "1", count: 1 }, { value: "2", count: 1 }]
};

it("renders correctly when a value is selected", () => {
  const wrapper = shallow(<Facet {...requiredProps} value="value" />);
  expect(wrapper).toMatchSnapshot();
});

it("renders correctly when a value is not selected", () => {
  const wrapper = shallow(<Facet {...requiredProps} />);
  expect(wrapper).toMatchSnapshot();
});
