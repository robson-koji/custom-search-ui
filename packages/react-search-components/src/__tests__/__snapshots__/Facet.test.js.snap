// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when a value is not selected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Facet
    name="Facet"
    onRemove={[Function]}
    onSelect={[Function]}
    options={
      Array [
        Object {
          "count": 1,
          "value": "1",
        },
        Object {
          "count": 1,
          "value": "2",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div>
        <div
          className="facet__title"
        >
          Facet
        </div>
        <ul
          className="facet__list"
        >
          <li
            className="facet__item"
          >
            <a
              className="facet__link"
              href="/"
              onClick={[Function]}
            >
              1
            </a>
             
            <span
              className="facet__count"
            >
              1
            </span>
          </li>
          <li
            className="facet__item"
          >
            <a
              className="facet__link"
              href="/"
              onClick={[Function]}
            >
              2
            </a>
             
            <span
              className="facet__count"
            >
              1
            </span>
          </li>
        </ul>
      </div>,
      "className": "facet",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="facet__title"
          >
            Facet
          </div>,
          <ul
            className="facet__list"
          >
            <li
              className="facet__item"
            >
              <a
                className="facet__link"
                href="/"
                onClick={[Function]}
              >
                1
              </a>
               
              <span
                className="facet__count"
              >
                1
              </span>
            </li>
            <li
              className="facet__item"
            >
              <a
                className="facet__link"
                href="/"
                onClick={[Function]}
              >
                2
              </a>
               
              <span
                className="facet__count"
              >
                1
              </span>
            </li>
          </ul>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Facet",
            "className": "facet__title",
          },
          "ref": null,
          "rendered": "Facet",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              undefined,
              Array [
                <li
                  className="facet__item"
                >
                  <a
                    className="facet__link"
                    href="/"
                    onClick={[Function]}
                  >
                    1
                  </a>
                   
                  <span
                    className="facet__count"
                  >
                    1
                  </span>
                </li>,
                <li
                  className="facet__item"
                >
                  <a
                    className="facet__link"
                    href="/"
                    onClick={[Function]}
                  >
                    2
                  </a>
                   
                  <span
                    className="facet__count"
                  >
                    1
                  </span>
                </li>,
              ],
            ],
            "className": "facet__list",
          },
          "ref": null,
          "rendered": Array [
            undefined,
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <a
                    className="facet__link"
                    href="/"
                    onClick={[Function]}
                  >
                    1
                  </a>,
                  " ",
                  <span
                    className="facet__count"
                  >
                    1
                  </span>,
                ],
                "className": "facet__item",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "1",
                    "className": "facet__link",
                    "href": "/",
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": "1",
                  "type": "a",
                },
                " ",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": 1,
                    "className": "facet__count",
                  },
                  "ref": null,
                  "rendered": 1,
                  "type": "span",
                },
              ],
              "type": "li",
            },
            Object {
              "instance": null,
              "key": "2",
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <a
                    className="facet__link"
                    href="/"
                    onClick={[Function]}
                  >
                    2
                  </a>,
                  " ",
                  <span
                    className="facet__count"
                  >
                    1
                  </span>,
                ],
                "className": "facet__item",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "2",
                    "className": "facet__link",
                    "href": "/",
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": "2",
                  "type": "a",
                },
                " ",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": 1,
                    "className": "facet__count",
                  },
                  "ref": null,
                  "rendered": 1,
                  "type": "span",
                },
              ],
              "type": "li",
            },
          ],
          "type": "ul",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div>
          <div
            className="facet__title"
          >
            Facet
          </div>
          <ul
            className="facet__list"
          >
            <li
              className="facet__item"
            >
              <a
                className="facet__link"
                href="/"
                onClick={[Function]}
              >
                1
              </a>
               
              <span
                className="facet__count"
              >
                1
              </span>
            </li>
            <li
              className="facet__item"
            >
              <a
                className="facet__link"
                href="/"
                onClick={[Function]}
              >
                2
              </a>
               
              <span
                className="facet__count"
              >
                1
              </span>
            </li>
          </ul>
        </div>,
        "className": "facet",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="facet__title"
            >
              Facet
            </div>,
            <ul
              className="facet__list"
            >
              <li
                className="facet__item"
              >
                <a
                  className="facet__link"
                  href="/"
                  onClick={[Function]}
                >
                  1
                </a>
                 
                <span
                  className="facet__count"
                >
                  1
                </span>
              </li>
              <li
                className="facet__item"
              >
                <a
                  className="facet__link"
                  href="/"
                  onClick={[Function]}
                >
                  2
                </a>
                 
                <span
                  className="facet__count"
                >
                  1
                </span>
              </li>
            </ul>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Facet",
              "className": "facet__title",
            },
            "ref": null,
            "rendered": "Facet",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                undefined,
                Array [
                  <li
                    className="facet__item"
                  >
                    <a
                      className="facet__link"
                      href="/"
                      onClick={[Function]}
                    >
                      1
                    </a>
                     
                    <span
                      className="facet__count"
                    >
                      1
                    </span>
                  </li>,
                  <li
                    className="facet__item"
                  >
                    <a
                      className="facet__link"
                      href="/"
                      onClick={[Function]}
                    >
                      2
                    </a>
                     
                    <span
                      className="facet__count"
                    >
                      1
                    </span>
                  </li>,
                ],
              ],
              "className": "facet__list",
            },
            "ref": null,
            "rendered": Array [
              undefined,
              Object {
                "instance": null,
                "key": "1",
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <a
                      className="facet__link"
                      href="/"
                      onClick={[Function]}
                    >
                      1
                    </a>,
                    " ",
                    <span
                      className="facet__count"
                    >
                      1
                    </span>,
                  ],
                  "className": "facet__item",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "1",
                      "className": "facet__link",
                      "href": "/",
                      "onClick": [Function],
                    },
                    "ref": null,
                    "rendered": "1",
                    "type": "a",
                  },
                  " ",
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": 1,
                      "className": "facet__count",
                    },
                    "ref": null,
                    "rendered": 1,
                    "type": "span",
                  },
                ],
                "type": "li",
              },
              Object {
                "instance": null,
                "key": "2",
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <a
                      className="facet__link"
                      href="/"
                      onClick={[Function]}
                    >
                      2
                    </a>,
                    " ",
                    <span
                      className="facet__count"
                    >
                      1
                    </span>,
                  ],
                  "className": "facet__item",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "2",
                      "className": "facet__link",
                      "href": "/",
                      "onClick": [Function],
                    },
                    "ref": null,
                    "rendered": "2",
                    "type": "a",
                  },
                  " ",
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": 1,
                      "className": "facet__count",
                    },
                    "ref": null,
                    "rendered": 1,
                    "type": "span",
                  },
                ],
                "type": "li",
              },
            ],
            "type": "ul",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`renders correctly when a value is selected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Facet
    name="Facet"
    onRemove={[Function]}
    onSelect={[Function]}
    options={
      Array [
        Object {
          "count": 1,
          "value": "1",
        },
        Object {
          "count": 1,
          "value": "2",
        },
      ]
    }
    value="value"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div>
        <div
          className="facet__title"
        >
          Facet
        </div>
        <ul
          className="facet__list"
        >
          <li
            className="facet__selected"
          >
            value
             
            <span
              className="facet__remove"
            >
              (
              <a
                href="/"
                onClick={[Function]}
              >
                Remove
              </a>
              )
            </span>
          </li>
        </ul>
      </div>,
      "className": "facet",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="facet__title"
          >
            Facet
          </div>,
          <ul
            className="facet__list"
          >
            <li
              className="facet__selected"
            >
              value
               
              <span
                className="facet__remove"
              >
                (
                <a
                  href="/"
                  onClick={[Function]}
                >
                  Remove
                </a>
                )
              </span>
            </li>
          </ul>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Facet",
            "className": "facet__title",
          },
          "ref": null,
          "rendered": "Facet",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <li
                className="facet__selected"
              >
                value
                 
                <span
                  className="facet__remove"
                >
                  (
                  <a
                    href="/"
                    onClick={[Function]}
                  >
                    Remove
                  </a>
                  )
                </span>
              </li>,
              false,
            ],
            "className": "facet__list",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "value",
                  " ",
                  <span
                    className="facet__remove"
                  >
                    (
                    <a
                      href="/"
                      onClick={[Function]}
                    >
                      Remove
                    </a>
                    )
                  </span>,
                ],
                "className": "facet__selected",
              },
              "ref": null,
              "rendered": Array [
                "value",
                " ",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      "(",
                      <a
                        href="/"
                        onClick={[Function]}
                      >
                        Remove
                      </a>,
                      ")",
                    ],
                    "className": "facet__remove",
                  },
                  "ref": null,
                  "rendered": Array [
                    "(",
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "Remove",
                        "href": "/",
                        "onClick": [Function],
                      },
                      "ref": null,
                      "rendered": "Remove",
                      "type": "a",
                    },
                    ")",
                  ],
                  "type": "span",
                },
              ],
              "type": "li",
            },
            false,
          ],
          "type": "ul",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div>
          <div
            className="facet__title"
          >
            Facet
          </div>
          <ul
            className="facet__list"
          >
            <li
              className="facet__selected"
            >
              value
               
              <span
                className="facet__remove"
              >
                (
                <a
                  href="/"
                  onClick={[Function]}
                >
                  Remove
                </a>
                )
              </span>
            </li>
          </ul>
        </div>,
        "className": "facet",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="facet__title"
            >
              Facet
            </div>,
            <ul
              className="facet__list"
            >
              <li
                className="facet__selected"
              >
                value
                 
                <span
                  className="facet__remove"
                >
                  (
                  <a
                    href="/"
                    onClick={[Function]}
                  >
                    Remove
                  </a>
                  )
                </span>
              </li>
            </ul>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Facet",
              "className": "facet__title",
            },
            "ref": null,
            "rendered": "Facet",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <li
                  className="facet__selected"
                >
                  value
                   
                  <span
                    className="facet__remove"
                  >
                    (
                    <a
                      href="/"
                      onClick={[Function]}
                    >
                      Remove
                    </a>
                    )
                  </span>
                </li>,
                false,
              ],
              "className": "facet__list",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "value",
                    " ",
                    <span
                      className="facet__remove"
                    >
                      (
                      <a
                        href="/"
                        onClick={[Function]}
                      >
                        Remove
                      </a>
                      )
                    </span>,
                  ],
                  "className": "facet__selected",
                },
                "ref": null,
                "rendered": Array [
                  "value",
                  " ",
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": Array [
                        "(",
                        <a
                          href="/"
                          onClick={[Function]}
                        >
                          Remove
                        </a>,
                        ")",
                      ],
                      "className": "facet__remove",
                    },
                    "ref": null,
                    "rendered": Array [
                      "(",
                      Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "host",
                        "props": Object {
                          "children": "Remove",
                          "href": "/",
                          "onClick": [Function],
                        },
                        "ref": null,
                        "rendered": "Remove",
                        "type": "a",
                      },
                      ")",
                    ],
                    "type": "span",
                  },
                ],
                "type": "li",
              },
              false,
            ],
            "type": "ul",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
